pingcap/tiflash
# KVStore: Support multi file snapshot

### What problem does this PR solve?

Issue Number: close #6086

Problem Summary:

see issue

### What is changed and how it works?

### Check List

Tests 


Integration test:

Whether and to what extent this case is covered by integration test is totally up to TiKV.

Currently, integration test can seldomly cover this case, which is caused by bugs on TiKV's side. See https://github.com/pingcap/tiflash/issues/5813.

Once TiKV loose its constraint, integration test will cover this case.

Manual test:
1. disable TiKV's limitation
2. create a playground with new TiKV and TiFlash
3. create a table only on TiKV
4. write many rows into this table
5. alter table set tiflash replica 1
6. wait this table until it becomes available

Found we can handle snapshot:
on Proxy
```
[2022/10/08 16:10:11.423 +08:00] [INFO] [observer.rs:700] ["pre apply snapshot"] [pending=0] [snap_key="SnapKey { region_id: 1005, term: 6, idx: 7 }"] [region_id=1005] [peer_id=1007]
[2022/10/08 16:10:11.424 +08:00] [INFO] [observer.rs:654] ["observe multi-file snapshot"] [total=2] [cf="\"default\""] [snap="Snapshot { key: SnapKey { region_id: 1005, term: 6, idx: 7 }, display_path: \"/Users/calvin/.tiup/data/multi/tiflash-0/flash/snap/rev_1005_6_7_(default|lock|write).sst\" }"]
[2022/10/08 16:10:11.616 +08:00] [INFO] [region.rs:439] ["begin apply snap data"] [peer_id=1007] [region_id=1005]
[2022/10/08 16:10:11.653 +08:00] [INFO] [snap.rs:670] ["set_snapshot_meta total cf files count: 4"]
[2022/10/08 16:10:11.792 +08:00] [INFO] [io.rs:259] ["apply_sst_cf_file starts on cf default. All files [\"/Users/calvin/.tiup/data/multi/tiflash-0/flash/snap/rev_1005_6_7_default.sst.clone\", \"/Users/calvin/.tiup/data/multi/tiflash-0/flash/snap/rev_1005_6_7_default_0001.sst.clone\"]"]
[2022/10/08 16:10:11.793 +08:00] [INFO] [observer.rs:767] ["post apply snapshot"] [region="id: 1005 start_key: 7480000000000000FF4B5F720000000000FA end_key: 7480000000000000FF4C00000000000000F8 region_epoch { conf_ver: 2 version: 37 } peers { id: 1006 store_id: 1 } peers { id: 1007 store_id: 70 role: Learner }"] [snap_key="SnapKey { region_id: 1005, term: 6, idx: 7 }"] [peer_id=1007]
[2022/10/08 16:10:11.922 +08:00] [INFO] [observer.rs:785] ["get prehandled snapshot success"] [pending=1] [region="id: 1005 start_key: 7480000000000000FF4B5F720000000000FA end_key: 7480000000000000FF4C00000000000000F8 region_epoch { conf_ver: 2 version: 37 } peers { id: 1006 store_id: 1 } peers { id: 1007 store_id: 70 role: Learner }"] [peer_id="SnapKey { region_id: 1005, term: 6, idx: 7 }"]
```

... and on TiFlash
```
[2022/10/08 16:10:11.440 +08:00] [INFO] [SSTFilesToBlockInputStream.cpp:84] ["SSTFilesToBlockInputStream:Read SST Files with MultiSSTReader"] [thread_id=12]
[2022/10/08 16:10:11.440 +08:00] [INFO] [SSTReader.h:93] ["MultiSSTReader:Init with /Users/calvin/.tiup/data/multi/tiflash-0/flash/snap/rev_1005_6_7_default.sstþ\u0007\u0005"] [thread_id=12]
[2022/10/08 16:10:11.440 +08:00] [INFO] [TiDBSchemaSyncer.h:113] ["SchemaSyncer:Start to sync schemas. current version is: 40 and try to sync schema version to: 41"] [thread_id=13]
[2022/10/08 16:10:11.440 +08:00] [DEBUG] [TiDBSchemaSyncer.h:170] ["SchemaSyncer:Try load schema diffs."] [thread_id=13]
[2022/10/08 16:10:11.441 +08:00] [INFO] [SSTReader.h:93] ["MultiSSTReader:Init with /Users/calvin/.tiup/data/multi/tiflash-0/flash/snap/rev_1005_6_7_write.sst\tâ\u0005"] [thread_id=12]
[2022/10/08 16:10:11.441 +08:00] [INFO] [SSTFilesToBlockInputStream.cpp:131] ["SSTFilesToBlockInputStream:Finish Construct MultiSSTReader, write 1 lock 0 default 2"] [thread_id=12]
[2022/10/08 16:10:11.441 +08:00] [DEBUG] [TiDBSchemaSyncer.h:181] ["SchemaSyncer:End load schema diffs with total 1 entries."] [thread_id=13]
[2022/10/08 16:10:11.442 +08:00] [DEBUG] [SchemaGetter.cpp:234] ["SchemaGetter:Get DB Info from TiKV : {\"id\":73,\"db_name\":{\"O\":\"test99\",\"L\":\"test99\"},\"charset\":\"utf8mb4\",\"collate\":\"utf8mb4_bin\",\"state\":5,\"policy_ref_info\":null}"] [thread_id=13]

...

[2022/10/08 16:10:11.622 +08:00] [INFO] [SSTReader.h:78] ["MultiSSTReader:Swithch to /Users/calvin/.tiup/data/multi/tiflash-0/flash/snap/rev_1005_6_7_default_0001.sst).sst\" }\u0004"] [thread_id=12]
[2022/10/08 16:10:11.686 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:251] ["SSTFilesToBlockInputStream:Done loading all kvpairs from [CF=default] [offset=1000] [write_cf_offset=1000] "] [thread_id=12]
[2022/10/08 16:10:11.694 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:251] ["SSTFilesToBlockInputStream:Done loading all kvpairs from [CF=lock] [offset=0] [write_cf_offset=1000] "] [thread_id=12]
[2022/10/08 16:10:11.806 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:251] ["SSTFilesToBlockInputStream:Done loading all kvpairs from [CF=default] [offset=1000] [write_cf_offset=1000] "] [thread_id=12]
[2022/10/08 16:10:11.806 +08:00] [DEBUG] [SSTFilesToBlockInputStream.cpp:251] ["SSTFilesToBlockInputStream:Done loading all kvpairs from [CF=lock] [offset=0] [write_cf_offset=1000] "] [thread_id=12]
[2022/10/08 16:10:11.810 +08:00] [DEBUG] [SSTFilesToDTFilesOutputStream.cpp:152] ["SSTFilesToDTFilesOutputStream:Create new DTFile for snapshot data, region=[region 1005, applied: term 6 index 7] file_idx=0 file=/Users/calvin/.tiup/data/multi/tiflash-0/data/data/t_75/stable/.tmp.dmf_2"] [thread_id=12]
```

We found default has 2 files, and MultiSSTReader can switch to the second file when finish reading the first one. The whole table can be available finally.

Side effects


Documentation


### Release note



```release-note
None
```
